icu4c/source/i18n/calendar.cpp

```diff
@@ -855,10 +855,19 @@ Calendar::createInstance(const TimeZone& zone, UErrorCode& success)
 
 // -------------------------------------
 
+#define ZID_KEY_MAX  128
 Calendar* U_EXPORT2
 Calendar::createInstance(const Locale& aLocale, UErrorCode& success)
 {
-    return createInstance(TimeZone::createDefault(), aLocale, success);
+    // Try to get @timezone= keyword from the locale. If exist, use that to
+    // create timezone. Otherwise, use the default.
+    char timezone[ZID_KEY_MAX];
+    UErrorCode tmpStatus = U_ZERO_ERROR;
+    int32_t len = aLocale.getKeywordValue("timezone", timezone, ZID_KEY_MAX, tmpStatus);
```

Comment:
Don't we have a version now that lets us get the value in a CharStringByteSink?

Reply:
Yes, there is:

http://icu-project.org/apiref/icu4c/classicu_1_1Locale.html#aef599483560e19d4f0ac8a65e0a4ddd4

And there is also a function that could return the value directly as an icu::CharString:

http://icu-project.org/apiref/icu4c/classicu_1_1Locale.html#abb04894db45b909d9f064e51a4d8b317